src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs

```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
                 });
         }
 
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ServerCallbackAndLocalCertificateSelectionSet_Throws(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                Assert.True(false, "LocalCertificateSelectionCallback called when AuthenticateAsServerAsync was expected to fail.");
+                return null;
+            });
+
+            var validationCallback = new RemoteCertificateValidationCallback((object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+            {
+                Assert.Equal(serverCert, certificate);
+                return true; 
+            });
+
+            VirtualNetwork vn = new VirtualNetwork();
+            using (VirtualNetworkStream serverStream = new VirtualNetworkStream(vn, isServer: true),
+                                        clientStream = new VirtualNetworkStream(vn, isServer: false))
+            using (SslStream server = new SslStream(serverStream, false, null, selectionCallback),
+                             client = new SslStream(clientStream, leaveInnerStreamOpen: false, validationCallback))
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                    Assert.True(false, "RemoteCertificateValidationCallback called when AuthenticateAsServerAsync was expected to fail.");
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
+                {
+                    timesCallbackCalled++;
+                    Assert.Equal(hostName, actualHostName);
+                    return serverCert;
+                };
+
+                var cts = new CancellationTokenSource();
+                Assert.Throws<InvalidOperationException>(() => {server.AuthenticateAsServerAsync(options, cts.Token).Wait();});
```

Comment:
Couple of things:
1) The CTS never has cancellation requested, so you can just pass in CancellationToken.None instead of creating a CTS and passing in its Token.
2) I'd prefer we avoid doing synchronous blocking like this.  Please change it to something like:
```C#
await Assert.ThrowsAsync<InvalidOperationException(() => server.AuthenticateAsServerAsync(options, CancellationToken.None));
```

Reply:
Thanks for the feedback. The other tests in the file are using the same pattern, so I'll update them all.

Reply:
I've modified all of these tests to avoid synchronous blocking. Feel free to point out anything I missed, or that I made async unnecessarily.